### DISCUSSION TITLE
Tables with autogenerated chars

### DISCUSSION THREAD

**Discussion by DonDom86** [EvotecIT/PSWriteHTML#342] [Ideas]
I love PSWriteHTML and the easy way to create HTML pages.
However, from my point of view, tables alone often look confusing.
The possibility of linking graphs and tables gave me an idea.

The function "Get-HTMLChart-Section" creates charts for a list of properties. 
Certainly not the "cleanest" code, but it is functional :-)

![image](https://user-images.githubusercontent.com/97181496/196521800-a45c9a3b-6640-4ab5-b15e-93f99c5b340c.png)

For linking charts and tables I need helper properties. Only strings ca be linked...right?
Is it possible to "hide" individual attributes of a row?
The same functionality as with responsive display of tables.
In my example "Enabled-String, DateFrom-Day, DateTo-Year and Date-Test-Month"

Like that...
![image](https://user-images.githubusercontent.com/97181496/196520998-11caa885-862e-4a26-8df2-0ce3c116acb2.png)

Best regards
Dominik

`Import-Module .\PSWriteHTML.psd1 -Force

#Define data
$DataTable = @(
    [PSCustomObject] @{
        Name     = 'My Object 1'
        Enabled  = $true
        Time     = 1
        TT       = 5
        AA       = 1
        DateFrom = (Get-Date).AddDays(1)
        DateTo   = (Get-Date)
        DateTest = (Get-Date).AddDays(0)
    }
    [PSCustomObject] @{
        Name     = 'My Object 2'
        Enabled  = $true
        Time     = "r"
        TT       = "aGDFGDFG"
        AA       = "4"
        DateFrom = (Get-Date).AddDays(2)
        DateTo   = (Get-Date).AddDays(3)
        DateTest = (Get-Date).AddDays(32)
    }
    [PSCustomObject] @{
        Name     = 'My Object 3'
        Enabled  = $true
        Time     = "a"
        TT       = "aDFGDFGsd"
        AA       = "daDFGDFGs"
        DateFrom = (Get-Date).AddDays(4)
        DateTo   = (Get-Date).AddDays(71)
        DateTest = (Get-Date).AddDays(64)
    }
    [PSCustomObject] @{
        Name     = 'My Object 3'
        Enabled  = $true
        Time     = "a"
        TT       = "aDFGDFGsd"
        AA       = "daDFGDFGs"
        DateFrom = (Get-Date).AddDays(5)
        DateTo   = (Get-Date).AddDays(7)
        DateTest = (Get-Date).AddDays(64)
    }
    [PSCustomObject] @{
        Name     = 'My Object 3'
        Enabled  = $true
        Time     = "a"
        TT       = "aDFGDdddFGsd"
        AA       = "daDFvvvGDFGs"
        DateFrom = (Get-Date).AddDays(6)
        DateTo   = (Get-Date).AddDays(7)
        DateTest = (Get-Date).AddDays(64)
    }
    [PSCustomObject] @{
        Name     = 'My Object 3'
        Enabled  = $true
        Time     = "aaa"
        TT       = "aDFGDddFGsd"
        AA       = "daDFGdasdDFGs"
        DateFrom = (Get-Date).AddDays(7)
        DateTo   = (Get-Date).AddDays(117)
        DateTest = (Get-Date).AddDays(64)
    }
    [PSCustomObject] @{
        Name     = 'My Object 3'
        Enabled  = $true
        Time     = "asada"
        TT       = "aDFGsadDFGsd"
        AA       = "daDFdasdGDFGs"
        DateFrom = (Get-Date).AddDays(9)
        DateTo   = (Get-Date).AddDays(7)
        DateTest = (Get-Date).AddDays(64)
    }
    [PSCustomObject] @{
        Name     = 'My Object 3'
        Enabled  = $false
        Time     = "aas"
        TT       = "aDFsssGDFGsd"
        AA       = "daDsadFGDFGs"
        DateFrom = (Get-Date).AddDays(9)
        DateTo   = (Get-Date).AddDays(7)
        DateTest = (Get-Date).AddDays(64)
    }
)

function Get-HTMLChart-Section {
    param(
        [Parameter(Position = 0, mandatory = $true)]
        [Object] $DataTable,
        [Parameter(Position = 0, mandatory = $true)]
        [string] $DataTableID,
        [Parameter(Position = 0, mandatory = $true)]
        $PropertyList,
        [int] $Min = 0,
        [int] $Max = [int]::PositiveInfinity
    )

    if ($null -ne $DataTable) {
        $PropertyNameTableList = $DataTable | Select-Object * -First 1 | ForEach-Object { $_.psobject.properties.name } | Select-Object -Unique
        $PropertyTableList = $DataTable | Select-Object * -First 1 | ForEach-Object { $_.psobject.properties } | Select-Object -Unique

        $PropertyNameMatchList = $PropertyNameTableList | select-string -pattern $PropertyList
        $Height = 250
        $Width = 250

        $NewHTMLSection = New-HTMLSection -Invisible -BorderRadius 0px -JustifyContent center { 
        
            foreach ($PropertyName in $PropertyNameMatchList) {

                $Property = $PropertyTableList | where-object { $_.Name -eq $PropertyName } | Select-Object * -First 1

                if ($Property.TypeNameOfValue -eq "System.DateTime") {
                    #yyyy
                    $GroupObjectyyyyList = $DataTable | Group-Object { $_.$($Property.Name).toString("yyyy") } | Sort-Object Name
                    $GroupObjectyyyyMMList = $DataTable | Group-Object { $_.$($Property.Name).toString("yyyy.MM") } | Sort-Object Name
                    $GroupObjectyyyyMMddList = $DataTable | Group-Object { $_.$($Property.Name).toString("yyyy.MM.dd") } | Sort-Object Name
                    if ($($GroupObjectyyyyList | Measure-Object | Select-Object * -ExpandProperty Count) -gt 1) {
                        $GroupObjectList = $GroupObjectyyyyList
                        $PropertyNamestring = $($Property.Name + "-" + "Year")
                        $DataTable | ForEach-Object { Add-Member -InputObject $_ -Membertype NoteProperty -Name $PropertyNamestring -value $_.$($Property.Name).toString("yyyy") }
                    }
                    #MM
                    elseIf ($($GroupObjectyyyyMMList | Measure-Object | Select-Object * -ExpandProperty Count) -gt 1) {
                        $GroupObjectList = $GroupObjectyyyyMMList
                        $PropertyNamestring = $($Property.Name + "-" + "Month")
                        $DataTable | ForEach-Object { Add-Member -InputObject $_ -Membertype NoteProperty -Name $PropertyNamestring -value $_.$($Property.Name).toString("yyyy.MM") }
                    }
                    #dd
                    elseIf ($($GroupObjectyyyyMMddList | Measure-Object | Select-Object * -ExpandProperty Count) -gt 1) {
                        $GroupObjectList = $GroupObjectyyyyMMddList
                        $PropertyNamestring = $($Property.Name + "-" + "Day")
                        $DataTable | ForEach-Object { Add-Member -InputObject $_ -Membertype NoteProperty -Name $PropertyNamestring -value $_.$($Property.Name).toString("yyyy.MM.dd") }
                    }
                    else {
                        $GroupObjectList = New-Object Collections.Generic.List[Object]
                    }
                    if ($GroupObjectList.Count -ge $Min -and $GroupObjectList.Count -le $Max ) {
                        New-HTMLSection -HeaderText $PropertyName.Pattern -Height $Height -Width $Width -BorderRadius 0px {
                            New-HTMLChart -Height $Height -Width $Width {
                                New-ChartBarOptions -Distributed -Type bar
                                foreach ($GroupObject in $GroupObjectList) {
                                    New-ChartBar -Name $GroupObject.Name -Value $GroupObject.Count
                                }
                                New-ChartLegend -HideLegend
                                $PropertyCount = $($DataTable | Select-Object * -First 1 | ForEach-Object { $_.psobject.properties.name } | Measure-Object | Select-Object * -Expand Count)
                                New-ChartEvent -DataTableID $DataTableID -ColumnID $($PropertyCount - 1)
                            }
                        }
                    }

                }
                else {
                    $PropertyNamestring = $($Property.Name + "-" + "String").Replace("\", "\\")
                    $DataTable | ForEach-Object { Add-Member -InputObject $_ -Membertype NoteProperty -Name $PropertyNamestring -value $_.$($Property.Name).toString() }

                    $GroupObjectList = $DataTable | Group-Object $PropertyNamestring
                    if ($GroupObjectList.Count -ge $Min -and $GroupObjectList.Count -le $Max ) {

                        New-HTMLSection -HeaderText $PropertyName.Pattern -Height $Height -Width $Width -BorderRadius 0px { 
                            New-HTMLChart -Height $Height -Width $Width {
                                foreach ($GroupObject in $GroupObjectList) {
                                    New-ChartPie -Name $GroupObject.Name -Value $GroupObject.Count 
                                }
                                New-ChartToolbar -Download 
                                New-ChartLegend -HideLegend
                                $PropertyCount = $($DataTable | Select-Object * -First 1 | ForEach-Object { $_.psobject.properties.name } | Measure-Object | Select-Object * -Expand Count)
                                New-ChartEvent -DataTableID $DataTableID -ColumnID $($PropertyCount - 1)
                            } 
                        } 

                    }
                }
            }
            
        }
    }
    return $NewHTMLSection
}

New-HTML {
    New-HTMLTableOption -DataStore JavaScript
    $PropertyList = @("Enabled", "DateFrom", "DateTo", "DateTest")
    Get-HTMLChart-Section -DataTable $DataTable -DataTableID $DataTable.GetHashCode() -PropertyList $PropertyList -Min 1 -Max 15
    New-HTMLTable -DataTable $DataTable -DataTableID $DataTable.GetHashCode()
} -ShowHTML -FilePath $PSScriptRoot\Example-ChartsWithTablesDonut.html -Online`

**Comment by PrzemyslawKlys** [comment]
I am not sure what you are asking for? Whether you can hide columns in tables? 

```powershell
$DataTable1 = @(
    [PSCustomObject] @{ Test = 'Name'; Test2 = 'Name2'; Test3 = 'Name3' }
    [PSCustomObject] @{ Test = 'Name'; Test2 = 'Name2'; Test3 = 'Name4' }
    [PSCustomObject] @{ Test = 'Name'; Test2 = 'Name2'; Test3 = 'Name5' }
    [PSCustomObject] @{ Test = 'Name'; Test2 = 'Name2'; Test3 = 'Name6' }
)

$DataTable1 | Out-HtmlView -Online -FilePath $PSScriptRoot\Example7_08_OutHTMlView.html -InvokeHTMLTags -SearchPane -Title 'Test' -DataStore JavaScript {
    New-TableColumnOption -ColumnIndex 1,2 -Hidden $true
    New-TableColumnOption -ColumnIndex 3 -Searchable $false
}
```

New-TableColumnOption can hide things and do other stuff

**Reply by DonDom86** [reply]
Thanks - working great and I think it also looks nice...
Maybe you can add a kind of this feature to a new version of PSWriteHTML? 
It provides a better overview and as a quick filter

![image](https://user-images.githubusercontent.com/97181496/196634090-8d7701fd-e6d9-459a-9553-e3e9f178d9ae.png)

```powershell
Import-Module .\PSWriteHTML.psd1 -Force

#Define data
$DataTable = @(
    [PSCustomObject] @{
        Name     = 'My Object 1'
        Enabled  = $true
        Time     = 1
        TT       = 5
        AA       = 1
        DateFrom = (Get-Date).AddDays(1)
        DateTo   = (Get-Date)
        DateTest = (Get-Date).AddDays(0)
    }
    [PSCustomObject] @{
        Name     = 'My Object 2'
        Enabled  = $true
        Time     = "r"
        TT       = "aGDFGDFG"
        AA       = "4"
        DateFrom = (Get-Date).AddDays(2)
        DateTo   = (Get-Date).AddDays(3)
        DateTest = (Get-Date).AddDays(32)
    }
    [PSCustomObject] @{
        Name     = 'My Object 3'
        Enabled  = $true
        Time     = "a"
        TT       = "aDFGDFGsd"
        AA       = "daDFGDFGs"
        DateFrom = (Get-Date).AddDays(4)
        DateTo   = (Get-Date).AddDays(71)
        DateTest = (Get-Date).AddDays(64)
    }
    [PSCustomObject] @{
        Name     = 'My Object 3'
        Enabled  = $true
        Time     = "a"
        TT       = "aDFGDFGsd"
        AA       = "daDFGDFGs"
        DateFrom = (Get-Date).AddDays(5)
        DateTo   = (Get-Date).AddDays(7)
        DateTest = (Get-Date).AddDays(64)
    }
    [PSCustomObject] @{
        Name     = 'My Object 3'
        Enabled  = $true
        Time     = "a"
        TT       = "aDFGDdddFGsd"
        AA       = "daDFvvvGDFGs"
        DateFrom = (Get-Date).AddDays(6)
        DateTo   = (Get-Date).AddDays(7)
        DateTest = (Get-Date).AddDays(64)
    }
    [PSCustomObject] @{
        Name     = 'My Object 3'
        Enabled  = $true
        Time     = "aaa"
        TT       = "aDFGDddFGsd"
        AA       = "daDFGdasdDFGs"
        DateFrom = (Get-Date).AddDays(7)
        DateTo   = (Get-Date).AddDays(117)
        DateTest = (Get-Date).AddDays(64)
    }
    [PSCustomObject] @{
        Name     = 'My Object 3'
        Enabled  = $true
        Time     = "asada"
        TT       = "aDFGsadDFGsd"
        AA       = "daDFdasdGDFGs"
        DateFrom = (Get-Date).AddDays(9)
        DateTo   = (Get-Date).AddDays(7)
        DateTest = (Get-Date).AddDays(64)
    }
    [PSCustomObject] @{
        Name     = 'My Object 3'
        Enabled  = $false
        Time     = "aas"
        TT       = "aDFsssGDFGsd"
        AA       = "daDsadFGDFGs"
        DateFrom = (Get-Date).AddDays(9)
        DateTo   = (Get-Date).AddDays(7)
        DateTest = (Get-Date).AddDays(64)
    }
)

function Get-HTMLChart-Section {
    param(
        [Parameter(Position = 0, mandatory = $true)]
        [Object] $DataTable,
        [Parameter(Position = 0, mandatory = $true)]
        [string] $DataTableID,
        [Parameter(Position = 0, mandatory = $true)]
        $PropertyList,
        [int] $Min = 0,
        [int] $Max = [int]::PositiveInfinity
    )

    $ColumIndexList = New-Object Collections.Generic.List[int]

    if ($null -ne $DataTable) {
        $PropertyNameTableList = $DataTable | Select-Object * -First 1 | ForEach-Object { $_.psobject.properties.name } | Select-Object -Unique
        $PropertyTableList = $DataTable | Select-Object * -First 1 | ForEach-Object { $_.psobject.properties } | Select-Object -Unique

        $PropertyNameMatchList = $PropertyNameTableList | select-string -pattern $PropertyList
        $Height = 250
        $Width = 250

        $NewHTMLSection = New-HTMLSection -Invisible -BorderRadius 0px -JustifyContent center { 
        
            foreach ($PropertyName in $PropertyNameMatchList) {

                $Property = $PropertyTableList | where-object { $_.Name -eq $PropertyName } | Select-Object * -First 1

                if ($Property.TypeNameOfValue -eq "System.DateTime") {
                    #yyyy
                    $GroupObjectyyyyList = $DataTable | Group-Object { $_.$($Property.Name).toString("yyyy") } | Sort-Object Name
                    $GroupObjectyyyyMMList = $DataTable | Group-Object { $_.$($Property.Name).toString("yyyy.MM") } | Sort-Object Name
                    $GroupObjectyyyyMMddList = $DataTable | Group-Object { $_.$($Property.Name).toString("yyyy.MM.dd") } | Sort-Object Name
                    if ($($GroupObjectyyyyList | Measure-Object | Select-Object * -ExpandProperty Count) -gt 1) {
                        $GroupObjectList = $GroupObjectyyyyList
                        $PropertyNamestring = $($Property.Name + "-" + "Year")
                        $DataTable | ForEach-Object { Add-Member -InputObject $_ -Membertype NoteProperty -Name $PropertyNamestring -value $_.$($Property.Name).toString("yyyy") }
                    }
                    #MM
                    elseIf ($($GroupObjectyyyyMMList | Measure-Object | Select-Object * -ExpandProperty Count) -gt 1) {
                        $GroupObjectList = $GroupObjectyyyyMMList
                        $PropertyNamestring = $($Property.Name + "-" + "Month")
                        $DataTable | ForEach-Object { Add-Member -InputObject $_ -Membertype NoteProperty -Name $PropertyNamestring -value $_.$($Property.Name).toString("yyyy.MM") }
                    }
                    #dd
                    elseIf ($($GroupObjectyyyyMMddList | Measure-Object | Select-Object * -ExpandProperty Count) -gt 1) {
                        $GroupObjectList = $GroupObjectyyyyMMddList
                        $PropertyNamestring = $($Property.Name + "-" + "Day")
                        $DataTable | ForEach-Object { Add-Member -InputObject $_ -Membertype NoteProperty -Name $PropertyNamestring -value $_.$($Property.Name).toString("yyyy.MM.dd") }
                    }
                    else {
                        $GroupObjectList = New-Object Collections.Generic.List[Object]
                    }
                    if ($GroupObjectList.Count -ge $Min -and $GroupObjectList.Count -le $Max ) {
                        New-HTMLSection -HeaderText $PropertyName.Pattern -Height $Height -Width $Width -BorderRadius 0px {
                            New-HTMLChart -Height $Height -Width $Width {
                                New-ChartBarOptions -Distributed -Type bar
                                foreach ($GroupObject in $GroupObjectList) {
                                    New-ChartBar -Name $GroupObject.Name -Value $GroupObject.Count
                                }
                                New-ChartLegend -HideLegend
                                $PropertyCount = $($DataTable | Select-Object * -First 1 | ForEach-Object { $_.psobject.properties.name } | Measure-Object | Select-Object * -Expand Count)
                                $ColumIndexList.add($PropertyCount - 1)
                                New-ChartEvent -DataTableID $DataTableID -ColumnID $($PropertyCount - 1)
                            }
                        }
                    }
                }
                else {
                    $PropertyNamestring = $($Property.Name + "-" + "String").Replace("\", "\\")
                    $DataTable | ForEach-Object { Add-Member -InputObject $_ -Membertype NoteProperty -Name $PropertyNamestring -value $_.$($Property.Name).toString() }

                    $GroupObjectList = $DataTable | Group-Object $PropertyNamestring
                    if ($GroupObjectList.Count -ge $Min -and $GroupObjectList.Count -le $Max ) {

                        New-HTMLSection -HeaderText $PropertyName.Pattern -Height $Height -Width $Width -BorderRadius 0px { 
                            New-HTMLChart -Height $Height -Width $Width {
                                foreach ($GroupObject in $GroupObjectList) {
                                    New-ChartPie -Name $GroupObject.Name -Value $GroupObject.Count 
                                }
                                New-ChartToolbar -Download 
                                New-ChartLegend -HideLegend
                                $PropertyCount = $($DataTable | Select-Object * -First 1 | ForEach-Object { $_.psobject.properties.name } | Measure-Object | Select-Object * -Expand Count)
                                $ColumIndexList.add($PropertyCount - 1)
                                New-ChartEvent -DataTableID $DataTableID -ColumnID $($PropertyCount - 1)
                            } 
                        } 
                    }
                }
            }
            
        }
    }

    $HTMLChartSectionObject = [PSCustomObject]@{
        NewHTMLSection = $NewHTMLSection
        ColumIndexList = $ColumIndexList

    }
    return $HTMLChartSectionObject
}


New-HTML {
    New-HTMLTableOption -DataStore JavaScript
    $PropertyList = @("Enabled", "DateFrom", "DateTo", "DateTest")

    $HTMLChartSectionObject = Get-HTMLChart-Section -DataTable $DataTable -DataTableID $DataTable.GetHashCode() -PropertyList $PropertyList -Min 1 -Max 15   

    New-HTMLContainer { $HTMLChartSectionObject.NewHTMLSection }

    New-HTMLTable -DataTable $DataTable -DataTableID $DataTable.GetHashCode() {
        New-TableColumnOption -ColumnIndex $HTMLChartSectionObject.ColumIndexList -Hidden $true

    }
} -ShowHTML -FilePath $PSScriptRoot\Example-ChartsWithTablesDonut.html -Online

``

**Reply by PrzemyslawKlys** [reply]
I am again not sure what you're asking - but if you mean modifying Out-HTMLView and adding charts to it by default it's not going to happen because it's impossible to automate it. While it may work for your case for small objects it's gonna be real issue for bigger tables.

**Comment by DonDom86** [comment]
Not per default :-) only as a fancy optional thing :-9
i have tested with Tables with 5000 Objects - not super fast, but it's okey... 
One click no creation of a filter needed.

Is it possible to deactivate red market scales?

![Ohne Titel 3](https://user-images.githubusercontent.com/97181496/196914163-1e92b6b8-e7a9-43ef-81a1-74e4790f9900.jpg)
![image](https://user-images.githubusercontent.com/97181496/196915617-8c99c529-82f4-4402-9c8b-2f283243160d.png)

Best regards
Dominik

**Comment by DonDom86** [comment]
Is it possible to deactivate red market scales?
